ledger/recentstorage/recentstorage_concrete.go

```diff
@@ -268,78 +271,142 @@ func (r *RecentIndexStorage) DecreaseIndexTTL(ctx context.Context) []core.Record
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	r.requests[obj] = &objContext
+	ctx = insmetrics.InsertTag(ctx, tagJet, r.jetID.DebugString())
+	stats.Record(ctx, statRecentStoragePendingsAdded.M(int64(len(objContext.Requests))))
+}
+
 // GetRequests returns a deep-copy of requests collections
-func (r *PendingStorage) GetRequests() map[core.RecordID]map[core.RecordID]struct{} {
+func (r *PendingStorageConcrete) GetRequests() map[core.RecordID]PendingObjectContext {
 	r.lock.RLock()
 	defer r.lock.RUnlock()
 
-	requestsClone := make(map[core.RecordID]map[core.RecordID]struct{})
-	for objID, objRequests := range r.requests {
-		objRequestsClone := make(map[core.RecordID]struct{}, len(objRequests))
-		for reqID, v := range objRequests {
-			objRequestsClone[reqID] = v
+	requestsClone := map[core.RecordID]PendingObjectContext{}
+	for objID, objContext := range r.requests {
+		objContext.lock.RLock()
+
+		objectClone := PendingObjectContext{
+			Active:   objContext.Active,
+			Requests: []core.RecordID{},
 		}
-		requestsClone[objID] = objRequestsClone
+		objectClone.Requests = append(objectClone.Requests, objContext.Requests...)
+		requestsClone[objID] = objectClone
+		objContext.lock.RUnlock()
 	}
 
 	return requestsClone
 }
 
 // GetRequestsForObject returns a deep-copy of requests collection for a specific object
-func (r *PendingStorage) GetRequestsForObject(obj core.RecordID) []core.RecordID {
+func (r *PendingStorageConcrete) GetRequestsForObject(obj core.RecordID) []core.RecordID {
 	r.lock.RLock()
 	defer r.lock.RUnlock()
 
 	forObject, ok := r.requests[obj]
 	if !ok {
 		return nil
 	}
-	results := make([]core.RecordID, 0, len(forObject))
-	for reqID := range forObject {
-		results = append(results, reqID)
-	}
+
+	forObject.lock.RLock()
+	defer forObject.lock.RUnlock()
+
+	results := make([]core.RecordID, 0, len(forObject.Requests))
+	results = append(results, forObject.Requests...)
 
 	return results
 }
 
 // RemovePendingRequest removes a request on object from cache
-func (r *PendingStorage) RemovePendingRequest(ctx context.Context, obj, req core.RecordID) {
-	r.lock.Lock()
-	defer r.lock.Unlock()
+func (r *PendingStorageConcrete) RemovePendingRequest(ctx context.Context, obj, req core.RecordID) {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
 
-	if _, ok := r.requests[obj]; !ok {
+	objContext, ok := r.requests[obj]
+	if !ok {
+		return
+	}
+
+	objContext.lock.Lock()
+	defer objContext.lock.Unlock()
+
+	firstRequest := objContext.Requests[0]
+	if firstRequest.Pulse() == req.Pulse() {
+		objContext.Active = true
+	}
+
+	index := -1
+	var objReq core.RecordID
+	for index, objReq = range objContext.Requests {
+		if objReq == req {
+			break
+		}
+	}
+
+	if index == -1 {
 		return
 	}
-	delete(r.requests[obj], req)
-	if len(r.requests[obj]) == 0 {
-		delete(r.requests, obj)
+
+	// edge-cases and a common path
+	// index == 0 - start of slice
+	// index == len - 1 - end of slice
+	// default - somewhere in slice
+	switch index {
```

Comment:
Let's check if `len(objContext.Requests) == 1` then remove the whole storage for the object (don't forget the write lock). After that we can only use the `default` case from the switch.